aiobotocore/endpoint.py

```diff
@@ -204,15 +204,9 @@ async def _do_get_response(self, request, operation_model, context):
         )
         parser = self._response_parser_factory.create_parser(protocol)
 
-        if asyncio.iscoroutinefunction(parser.parse):
-            parsed_response = await parser.parse(
-                response_dict, operation_model.output_shape
-            )
-        else:
-            parsed_response = parser.parse(
-                response_dict, operation_model.output_shape
-            )
-
+        parsed_response = await parser.parse(
+            response_dict, operation_model.output_shape
+        )
```

Comment:
have we validated across all the parsers this is now true?  Are there parsers that don't inherit from our baseclass?

Reply:
i see everything parents to ResponseParser 👍🏼

Reply:
Good call!

Upstream has a number of parsers that we have not patched:
* `BaseXMLResponseParser`
* `BaseJSONParser`
* `BaseEventStreamParser`
* `EventStreamJSONParser`
* `EventStreamXMLParser`
* `BaseRestParser`

The `Base*` parsers should be fine, since we patch any relevant child parsers. But I believe the `EventStream*` parsers might be problematic. They are used indirectly and might need to be patched to work correctly.

Reply:
I could prepare a follow-up of #1312 to fully patch the parser class hierarchy. That would be a prerequisite for this PR, IMO. Any thoughts?

Reply:
Work-in-progress: #1329

Reply:
#1329 makes sure that `EVENT_STREAM_PARSER_CLS` class attributes ("indirect" uses) point to patched parsers. Once that is merged, we can continue with this one.

Reply:
Interesting, I didn't look at `EVENT_STREAM_PARSER_CLS` and how they are used and inherited. If I understand them right, they are based on the sync implementation for parser. `aiobotocore` has no async version of them, probably because these event parsers don't do I/O, right? As far as I understand, they are used by `AioEventStream` to parse each received "chunk". My belief atm is that not having an async version for `EventStream*` is fine. 

Just the classes in the `PROTOCOL_PARSERS` mapping are changed here to provide a consistent async interface for the `parser` method. What would be the benefit of fully patching the parser class hierarchy?

Reply:
> My belief atm is that not having an async version for `EventStream*` is fine.

I guess it would be, as long as we continue to allow for both sync and async variants of `parse()`. However, this PR takes away that flexibility and we need to be careful to handle all parser classes, as @thehesiod rightfully pointed out. As far as I understand, that includes parser classes that we have not patched previously, hence #1329. I could well be on the wrong track here, though.

An alternative would be to simply keep the current implementation of `parse()` and move it up the `AioResponseParser`, but leave everything else as it is.

Reply:
ya that's why I was waiting to do a review of this PR when I had some time, we need to thoroughly look at how these classes are used

Reply:
lets report back when one of us does that

Reply:
How about allowing sync `parse` methods but emit a deprecation warning?

Reply:
sounds like a plan, would be safer